bootstrap/090_post_setup.sql

```diff
@@ -169,33 +169,117 @@ CREATE OR REPLACE TASK TASKS.COST_CONTROL_MONITORING
+    ON q.name = new_usage.name AND q.persona = new_usage.persona AND q.day = new_usage.day AND q.hour_of_day = new_usage.hour_of_day
+    WHEN MATCHED THEN
+        -- Set the new credits_used as we recomputed that whole hour
+        UPDATE SET q.credits_used = new_usage.credits_used, q.last_updated = :start_time
+    WHEN NOT MATCHED THEN
+        INSERT (name, credits_used, day, hour_of_day, persona, last_updated)
+        VALUES(new_usage.name, new_usage.credits_used, new_usage.day, new_usage.hour_of_day, new_usage.persona, :start_time);
+
+    -- Record the number of rows we wrote
+    let rows_added integer := SQLROWCOUNT;
+
+    -- Save the output of that query so we can return it from the task to ease debugging.
+    let quota_outcome string := 'Updated ' || :rows_added || ' rows in aggregated_hourly_quota since ' || :qh_func_start_time;
+
+    -- Aggregate the usage for today from aggregated_hourly_quota into the map that the external function expects.
+    -- e.g. {'users': {'bob': 1.0}, 'roles': {'PUBLIC': 2.0}}
+    let quota_usage object := (
+        with todays_usage as (
+            select * from internal.aggregated_hourly_quota
+            where day = TO_DATE(:start_time)
+        ), user_usage as(
+            -- Exclude the oldest hourly buckets of cached credit use data as we are getting fresh data
+            -- from the information_schema.query_history table function to avoid double-counting.
+            SELECT name, sum(credits_used) as credits
+            FROM todays_usage
+            WHERE persona = 'user'
+            group by name
+        ), role_usage as (
+            SELECT name, sum(credits_used) as credits
+            FROM todays_usage
+            WHERE persona = 'role'
+            group by name
+        ), aggr as (
+            select 'users' as kind, object_agg(name, credits) as usage_map from user_usage
+            union all
+            select 'roles' as kind, object_agg(name, credits) as usage_map from role_usage
+        )
+        select object_agg(kind, usage_map) from aggr);
+
+    -- Send the result to Sundeck, record the outcome from calling the external function
     BEGIN
-        quota_outcome := (select to_json(INTERNAL.REPORT_QUOTA_USED(:quota_usage)));
+        quota_outcome := (select :quota_outcome || '. ' || to_json(INTERNAL.REPORT_QUOTA_USED(:quota_usage)));
     EXCEPTION
         WHEN other THEN
             SYSTEM$LOG_ERROR(OBJECT_CONSTRUCT('error', 'External function to report cost usage failed.', 'SQLCODE', :sqlcode, 'SQLERRM', :sqlerrm, 'SQLSTATE', :sqlstate));
-            quota_outcome := :sqlerrm;
+            quota_outcome := :quota_outcome || '. Failed to report quota usage to Sundeck ' || :sqlerrm;
     END;
 
     -- Log the results locally
-    insert into internal.quota_task_history select :start_time, current_timestamp(), :quota_usage, :quota_outcome;
+    INSERT INTO internal.quota_task_history SELECT :start_time, current_timestamp(), :quota_usage, :quota_outcome;
```

Comment:
for some reason, `:quota_outcome` is always `null` now and I can't yet figure out why.

Reply:
`varchar || null` nets a `null`. Whoops.